### ISSUE TITLE
Add option to execute an arbitrary command instead of sending signal to process

### ISSUE THREAD

**Issue by x-cray** [hashicorp/envconsul#54]
There is a use case when process is controlled already by some kind of orchestrator. For example, Mesosphere Marathon which ensures service availability by controlling how many instances should be started and where. It would be beneficial if envconsul could delegate service instance restarting job to external tool by executing an arbitrary command. Such command might call Marathon API to do rolling restart of application instances while maintaining necessary health capacity.

**Comment by sethvargo** [comment]
Hi @x-cray 

You can read my comment on #52 for more information, but envconsul runs a child process, and once that child process is spawned, there are very few entry points besides signals. Even if you could specify an arbitrary command, envconsul would still need to kill and re-spawn the application. Sorry!

**Comment by x-cray** [comment]
@sethvargo Ok, I see your point. But this feature is not affected by the fact that it's impossible to replace environment. The process scheduler which I refer to actually takes care of running envconsul, so in my case I don't really need envconsul to manage my process. What I need from it is to get data from Consul, start child process with new environment and tell me when Consul data is changed so I can instruct main process scheduler to completely restart envconsul. If you're wiling to have such a feature in envconul I'll try to come up with a pull request.

**Comment by sethvargo** [comment]
@x-cray then why not just start the child process with bash and handle the `at_exit` with a trap?

**Comment by x-cray** [comment]
@sethvargo it makes sense, I've completely forgot that there is a `wait` parameter which may be set to arbitrary amount of time to let the bash script to call scheduler. Thanks

**Comment by x-cray** [comment]
Hi @sethvargo. So I tried to write a service wrapper like you suggested. For this scenario I needed to pass multiple signals to it:
* SIGUSR1 which triggers service restart by external scheduler, originated from envconsul in response to KV data change.
* SIGTERM which prepares service for graceful shutdown, originated from external scheduler.

With #73 I hoped to achieve the desired result but with current envconsul design it's impossible to send the second signal, there is an issue which prevents such use case, namely `Runner` won't pass any signals to the child process while `Runner.killProcess()` is waiting for it for `config.Timeout` seconds. The reason behind that is `Runner` synchronisation — `Runner.Signal()` attempts to acquire lock while it's being already acquired by `Runner.killProcess()` caller, so we're stuck in `Runner.Signal()` until the child process exits and eventually attempt to send the second signal to already-exited process.

Not sure whether such design was intentional, but I came up with a small [Bash service wrapper](https://github.com/x-cray/service-wrapper/blob/master/service-wrapper.sh) which implements Consul KV watcher by itself and works for me well.
